amazon-kinesis-client/src/main/java/software/amazon/kinesis/leases/LeaseCleanupManager.java

```diff
@@ -0,0 +1,299 @@
+    @NonNull
+    private final Duration maxFutureWait;
+    @NonNull
+    private final ScheduledExecutorService deletionThreadPool;
+    private final boolean cleanupLeasesUponShardCompletion;
+    private final long leaseCleanupIntervalMillis;
+    private final long completedLeaseCleanupThresholdMillis;
+    private final long garbageLeaseCleanupThresholdMillis;
+
+    private final Queue<LeasePendingDeletion> deletionQueue = new ConcurrentLinkedQueue<>();
+
+    private static final int MAX_RECORDS = 1;
+    private static final long INITIAL_DELAY = 0L;
+    private static final String LEASE_CLEANUP_MANAGER = "LeaseCleanupManager";
+
+    @Getter
+    private volatile boolean isRunning = false;
+
+    /**
+     * Starts the lease cleanup thread, which is scheduled periodically as specified by
+     * {@link LeaseCleanupManager#leaseCleanupIntervalMillis}
+     */
+    public void start() {
+        log.debug("Starting lease cleanup thread.");
+        deletionThreadPool.scheduleAtFixedRate(new LeaseCleanupThread(), INITIAL_DELAY, leaseCleanupIntervalMillis,
+                TimeUnit.MILLISECONDS);
+    }
+
+    /**
+     * Enqueues a lease for deletion.
+     * @param leasePendingDeletion
+     */
+    public void enqueueForDeletion(LeasePendingDeletion leasePendingDeletion) {
+        final Lease lease = leasePendingDeletion.lease();
+        if (lease == null) {
+            log.warn("Cannot enqueue lease {} for deferred deletion - instance doesn't hold the lease for that shard.",
+                    lease.leaseKey());
+        } else {
+            if (!deletionQueue.contains(leasePendingDeletion)) {
+                log.debug("Enqueuing lease {} for deferred deletion.", lease.leaseKey());
+                deletionQueue.add(leasePendingDeletion);
+            } else {
+                log.warn("Lease {} is already pending deletion, not enqueueing for deletion.", lease.leaseKey());
+            }
+        }
+    }
+
+    /**
+     * Returns how many leases are currently waiting in the queue pending deletion.
+     * @return number of leases pending deletion.
+     */
+    public int leasesPendingDeletion() {
+        return deletionQueue.size();
+    }
+
+    private boolean cleanupLease(LeasePendingDeletion leasePendingDeletion) throws InterruptedException, ExecutionException,
+            TimeoutException, DependencyException, ProvisionedThroughputException, InvalidStateException {
+        final Lease lease = leasePendingDeletion.lease();
+        final ShardInfo shardInfo = leasePendingDeletion.shardInfo();
+        final StreamIdentifier streamIdentifier = leasePendingDeletion.streamIdentifier();
+
+        final AWSExceptionManager exceptionManager = createExceptionManager();
+        boolean cleanedUpLease = false;
+        boolean alreadyCheckedForGarbageCollection = false;
+
+        try {
+            if (leasePendingDeletion.pendingForLongerThan(completedLeaseCleanupThresholdMillis)) {
```

Comment:
Hmmm this timer should not be based on the time spent by element in the queue, but rather based on the last scan for completed shard or garbage shard